File: 1586-longest-subarray-of-1s-after-deleting-one-element/longest-subarray-of-1s-after-deleting-one-element.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums) {
        int n = nums.size();
        int l = 0, r = 0;
        int zeros = 0;
        int maxlen = 0;

        while(r < n){
            if(nums[r] == 0) zeros++;

            while(zeros > 1){
                if(nums[l] == 0)
                zeros--;
                l++;
            }

            maxlen = max(maxlen,r-l+1);
            r++;
        }
        return maxlen - 1;
    }
};

File: 1586-longest-subarray-of-1s-after-deleting-one-element/longest-subarray-of-1s-after-deleting-one-element.java
class Solution {
    public int longestSubarray(int[] nums) {
        int left = 0, zeros = 0, right = 0, maxlen = 0;

        while(right < nums.length){
            if(nums[right] == 0) zeros++;

            while(zeros > 1) {
                if(nums[left] == 0) zeros--;
                left++;
            }

            maxlen = Math.max(maxlen,right - left); //most important -1 for deleting
            right++;
        }
        return maxlen;
    }
}